src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+   void *shm_start; // Points to start address of shared segment.
+   void *shm_end;   // Points to end address; one past end of sh segment
+   void *shm_next;  // Points to next 'free' address to allocate from.
+   void *shm_splinterdb_handle;
+   void *shm_large_frag_hip; // Highest addr of large-fragments tracked
+
+   platform_spinlock shm_mem_frags_lock;
+   // Protected by shm_mem_frags_lock. Must hold to read or modify.
+   shm_frag_info shm_mem_frags[SHM_NUM_LARGE_FRAGS];
+   uint32        shm_num_frags_tracked;
+   uint32        shm_num_frags_inuse;
+   uint32        shm_num_frags_inuse_HWM;
+   size_t        shm_used_by_large_frags; // Actually reserved
+   size_t        shm_nfrees;              // # of calls to free memory
+   size_t        shm_nf_search_skipped;   // See above;
+
+   size_t shm_total_bytes; // Total size of shared segment allocated initially.
+   size_t shm_free_bytes;  // Free bytes of memory left (that can be allocated)
+   size_t shm_used_bytes;  // Used bytes of memory left (that were allocated)
+   uint64 shm_magic;       // Magic identifier for shared memory segment
+   int    shm_id;          // Shared memory ID returned by shmget()
+
+} PLATFORM_CACHELINE_ALIGNED shmem_info;
+
+/* Permissions for accessing shared memory and IPC objects */
+#define PLATFORM_IPC_OBJS_PERMS (S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP)
+
+// Function prototypes
+
+static void
+platform_shm_track_alloc(shmem_info *shm, void *addr, size_t size);
+
+static void
+platform_shm_track_free(shmem_info *shm,
+                        void       *addr,
+                        const char *objname,
+                        const char *func,
+                        const char *file,
+                        const int   lineno);
+
+static void *
+platform_shm_find_free(shmem_info *shm,
+                       size_t      size,
+                       const char *objname,
+                       const char *func,
+                       const char *file,
+                       const int   lineno);
+
+static void
+platform_shm_trace_allocs(shmem_info  *shminfo,
+                          const size_t size,
+                          const char  *verb,
+                          const void  *retptr,
+                          const char  *objname,
+                          const char  *func,
+                          const char  *file,
+                          const int    lineno);
+
+static int
+platform_trace_large_frags(shmem_info *shm);
+
+/*
+ * PLATFORM_HEAP_ID_TO_HANDLE() --
+ * PLATFORM_HEAP_ID_TO_SHMADDR() --
+ *
+ * The shared memory create function returns the address of shmem_info->shm_id
+ * as the platform_heap_id heap-ID to the caller. Rest of Splinter will use this
+ * heap-ID as a 'handle' to manage / allocate shared memory. This macro converts
+ * the heap-ID handle to the shared memory's start address, from which the
+ * location of the next-free-byte can be tracked.
+ */
+static inline platform_heap_handle
```

Comment:
WHOLE bunch of conversion / access lookup functions. You can skip this pile.

Reply:
Deleted `platform_heap_handle *` from the code base and from this file. Reworked mapping interfaces to only use `platform_heap_id`.